src/libpmemset/part.c

```diff
@@ -159,35 +160,56 @@ pmemset_part_map_new(struct pmemset_part_map **part_map,
 	struct pmem2_source *pmem2_src;
 	pmem2_src = pmemset_file_get_pmem2_source(part->file);
 
-	/* mapping with a hint address next to previous mapped part */
-	void *contiguous_addr = (char *)previous_part.addr + previous_part.size;
 	size_t part_size = part->length;
 	if (part_size == 0)
 		pmem2_source_size(pmem2_src, &part_size);
 
 	struct pmem2_vm_reservation *pmem2_reserv;
-	pmem2_vm_reservation_new(&pmem2_reserv, contiguous_addr, part_size);
+	size_t offset = 0;
 
-	pmem2_config_set_vm_reservation(pmem2_cfg, pmem2_reserv, 0);
+	/* coalesce part mappings into single vm reservation */
+	if (part_coalescing && prev_pmap) {
+		pmem2_reserv = prev_pmap->pmem2_reserv;
+		offset = pmem2_vm_reservation_get_size(pmem2_reserv);
+		ret = pmem2_vm_reservation_extend(pmem2_reserv, part_size);
+	} else
+		ret = pmem2_vm_reservation_new(&pmem2_reserv, 0, part_size);
+
+	if (ret) {
+		if (ret == PMEM2_E_MAPPING_EXISTS)
+			ret = PMEMSET_E_CANNOT_COALESCE_PART_MAPS;
+		Free(pmap);
+		goto err_cfg_delete;
+	}
+
+	pmem2_config_set_vm_reservation(pmem2_cfg, pmem2_reserv, offset);
 
 	struct pmem2_map *pmem2_map;
 	ret = pmem2_map_new(&pmem2_map, pmem2_cfg, pmem2_src);
 	if (ret) {
 		ERR("cannot create pmem2 mapping %d", ret);
 		ret = PMEMSET_E_INVALID_PMEM2_MAP;
 		Free(pmap);
-		if (pmem2_reserv)
+		if (prev_pmap) {
+			size_t offset = pmem2_vm_reservation_get_size(
+					pmem2_reserv) - part_size;
+			pmem2_vm_reservation_shrink(pmem2_reserv, offset,
+					part_size);
+		} else
 			pmem2_vm_reservation_delete(&pmem2_reserv);
 		goto err_cfg_delete;
 	}
 
+	if (!(part_coalescing && prev_pmap)) {
```

Comment:
create a variable like this: `int coalesce = part_coalescing && prev_pmap;` and then use it in the two conditions you have.